File: fe1-web/src/model/network/method/message/data/chirp/NotifyDeleteChirp.ts
import { Hash, Timestamp } from 'model/objects';
import { ProtocolError } from 'model/network/ProtocolError';
import { validateDataObject } from 'model/network/validation';
import { checkTimestampStaleness } from '../Checker';
import { ActionType, MessageData, ObjectType } from '../MessageData';

/** Data sent to broadcast a chirp to the general social channel */
export class NotifyDeleteChirp implements MessageData {
  public readonly object: ObjectType = ObjectType.CHIRP;

  public readonly action: ActionType = ActionType.NOTIFY_DELETE;

  // The ID of the chirp
  public readonly chirp_id: Hash;

  // The channel to which the chirp has been deleted
  public readonly channel: string;

  // The timestamp at which the chirp has been deleted
  public readonly timestamp: Timestamp;

  constructor(msg: Partial<NotifyDeleteChirp>) {
    if (!msg.chirp_id) {
      throw new ProtocolError("Undefined 'id' parameter encountered during 'NotifyDeleteChirp'");
    }
    this.chirp_id = msg.chirp_id;

    if (!msg.timestamp) {
      throw new ProtocolError(
        "Undefined 'timestamp' parameter encountered during 'NotifyDeleteChirp'",
      );
    }
    checkTimestampStaleness(msg.timestamp);
    this.timestamp = msg.timestamp;

    if (!msg.channel) {
      throw new ProtocolError(
        "Undefined 'channel' parameter encountered during 'NotifyDeleteChirp'",
      );
    }
    this.channel = msg.channel;
  }

  /**
   * Creates a NotifyDeleteChirp object from a given JSON object.
   *
   * @param obj - The given JSON object
   */
  public static fromJson(obj: any): NotifyDeleteChirp {
    const { errors } = validateDataObject(ObjectType.CHIRP, ActionType.NOTIFY_DELETE, obj);

    if (errors !== null) {
      throw new ProtocolError(`Invalid receive chirp\n\n${errors}`);
    }

    return new NotifyDeleteChirp({
      ...obj,
      chirp_id: new Hash(obj.chirp_id),
      timestamp: new Timestamp(obj.timestamp),
    });
  }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/socialMedia/NotifyDeleteChirp.scala
package ch.epfl.pop.model.network.method.message.data.socialMedia

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.ActionType.ActionType
import ch.epfl.pop.model.network.method.message.data.ObjectType.ObjectType
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Channel, Hash, Timestamp}
import spray.json._

final case class NotifyDeleteChirp(
    chirp_id: Hash,
    channel: Channel,
    timestamp: Timestamp
) extends MessageData {
  override val _object: ObjectType = ObjectType.CHIRP
  override val action: ActionType = ActionType.NOTIFY_DELETE
}

object NotifyDeleteChirp extends Parsable {
  def apply(
      chirp_id: Hash,
      channel: Channel,
      timestamp: Timestamp
  ): NotifyDeleteChirp = new NotifyDeleteChirp(chirp_id, channel, timestamp)

  override def buildFromJson(payload: String): NotifyDeleteChirp = payload.parseJson.asJsObject.convertTo[NotifyDeleteChirp]
}
